src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs

```diff
@@ -681,22 +681,23 @@ private ModuleToken HandleToModuleToken(ref CORINFO_RESOLVED_TOKEN pResolvedToke
             {
                 object resultDef = methodILDef.GetObject((int)pResolvedToken.token);
 
-                if (resultDef is MethodDesc)
+                if (resultDef is MethodDesc resultMethod)
                 {
-                    if (resultDef is IL.Stubs.PInvokeTargetNativeMethod rawPinvoke)
-                        resultDef = rawPinvoke.Target;
-
-                    Debug.Assert(resultDef is EcmaMethod);
-                    Debug.Assert(_compilation.NodeFactory.CompilationModuleGroup.VersionsWithType(((EcmaMethod)resultDef).OwningType));
-                    token = (mdToken)MetadataTokens.GetToken(((EcmaMethod)resultDef).Handle);
-                    module = ((EcmaMethod)resultDef).Module;
+                    if (resultMethod is IL.Stubs.PInvokeTargetNativeMethod rawPinvoke)
+                        resultMethod = rawPinvoke.Target;
+                    resultMethod = resultMethod.GetTypicalMethodDefinition();
```

Comment:
Doesn't this lose information? If we're not recording a MethodSpec token for the instantiated method and strip the instantiation away, how will the runtime know what method instantiation to fix this up with at runtime? Wouldn't the generated fixup after this change say "call uninstantiated version of this method"? (That's why the comment above says `This only works for **non-generic types/members within the current version bubble**`.)

The problem is that we don't have a (guaranteed to exist) MethodSpec token for this signature in the compiled module because we just made this signature up. And since crossgen2 (or crossgen) can't make up new tokens, this ends up being not representable. We'll probably have to get into the token-making business at some point, but it kind of sounds like a big workitem.

I wonder if our only option here is to inline the generated IL for GetArrayDataReference into the marshalling stub because that one is not generic.

It's unfortunate because I was quite happy to see private APIs being replaced by public APIs in the generated stub.

Reply:
I have been thinking about this for about an hour while driving to Prague and back and I have managed to convince myself it's actually not a problem; but I agree that the reasoning behind this opinion is somewhat subtle and I'm ready to give in if you insist on inlining the helper in the marshalling code.

The trick is that the only purpose of ModuleToken's is for encoding method signatures in the import cells. As you can easily see at

https://github.com/dotnet/runtime/blob/c7f4ceb7f09148a9a365dad0f1174dd7da22cc69/src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/SignatureBuilder.cs#L475

we're always encoding instantiated methods by stripping the instantiation and encoding it explicitly in the signature - the switch at line 510 even explicitly asserts that at this point we're no longer working with a methodSpec.

Reply:
(the same holds for instantiated field signatures)

Reply:
Mmm, that is indeed subtle. But it is preferable for the generated marshallers to only call into public APIs (we might need to do that to ship) - so let's go with that.

Maybe delete the `non-generic` part of the comment above and add a short comment to each of the `GetTypical` calls that says something along the lines of `// It's okay to strip instantiation away because we don't need a TypeSpec token - we'll grab the instantiation from the type system entity later`

Reply:
Comment updated in 2nd commit.